macros/src/trans.rs

```diff
@@ -204,6 +213,46 @@ fn init(app: &App, main: &mut Vec<Tokens>, root: &mut Vec<Tokens>) {
         exprs.push(quote!(init::Resources::new()));
     }
 
+    let mut late_resources = vec![];
+    let has_late_resources = app.resources.iter()
+        .find(|&(_, res)| res.expr.is_none()).is_some();
+
+    if has_late_resources {
+        // `init` must initialize and return resources
+
+        let mut fields = vec![];
+
+        for (name, resource) in app.resources.iter()
+            .filter(|&(_, res)| res.expr.is_none()) {
+            let _name = Ident::new(format!("_{}", name.as_ref()));
+
+            let ty = &resource.ty;
+
+            fields.push(quote! {
+                pub #name: #ty,
+            });
+
+            late_resources.push(quote! {
+                #_name = #krate::UntaggedOption { some: _late_resources.#name };
```

Comment:
nit: I think `UntaggedOption::some(..)` would be shorter. Would that require the `const_fn` feature in user code though? Maybe it's already required by other generated code ...

Reply:
Actually, `#_name.some = _late_resources.#name` is even shorter. I'm not 100% if that has the exact semantics wrt to destructors. I think they are never run for any assignment.

Reply:
> Would that require the const_fn feature in user code though?

Yes. That's the reason why I made those fields public in the first place.

> Actually, #_name.some = _late_resources.#name is even shorter.

Yeah, I'll do this

> I'm not 100% if that has the exact semantics wrt to destructors. I think they are never run for any assignment.

Running them would be unsafe (how does the drop glue know which variant was active before the assignment?), so it can't really happen automatically (dropping or assignment is always safe). At least that's how I understand it.